src/cache.rs

```diff
@@ -5,130 +5,111 @@
-    pub(crate) fn persist(&self) -> Result<(), Error> {
-        let data_dir = CACHE_DATA_PATH.get().expect("cache path must be set");
-        let id_path = data_dir.join("cache.id");
-
-        let persistence_id = Persistence::persist(&backend(), &self.data)?;
-
-        let mut bytes = [0u8; 36];
-        let mut sink = Sink::new(&mut bytes);
-
-        persistence_id.encode(&mut sink);
+        let write_batch = WriteBatch::default();
 
-        fs::write(id_path, bytes)?;
-
-        Ok(())
+        Ok(Self { db, write_batch })
     }
 
-    /// Insert a note into the cache at the given block height. If note is
-    /// `None` only the block height gets updated.
+    // We store a column family named by hex representation of the psk.
+    // We store key as (psk, note) and value as (KeyData)
+    // The Canon of the tuple, we use it so we can have unique keys per psk and
+    // note
     pub(crate) fn insert(
         &mut self,
         psk: PublicSpendKey,
         height: u64,
         note: Option<Note>,
     ) -> Result<(), Error> {
-        if self.data.get(&psk)?.is_none() {
-            self.data.insert(
-                psk,
-                KeyData {
-                    last_height: 0,
-                    notes: BTreeSet::new(),
-                },
-            )?;
-        }
+        let cf: &ColumnFamily;
+        let cf_name = format!("{:?}", psk);
 
-        let mut key_data = self.data.get_mut(&psk)?.expect("psk");
+        let mut data = KeyData {
+            last_height: height,
+            notes: BTreeSet::new(),
+        };
+
+        if let Some(column_family) = self.db.cf_handle(&cf_name) {
+            cf = column_family;
+        } else {
+            // immediately create a cf by the hex of the psk_bytes
+            self.db.create_cf(&cf_name, &Options::default())?;
 
-        if height > key_data.last_height {
-            key_data.last_height = height;
+            cf = self
+                .db
+                .cf_handle(&cf_name)
+                .expect("cannot create column family for db");
+        }
+
+        if height > data.last_height {
+            data.last_height = height;
         }
 
         if let Some(note) = note {
-            key_data.notes.insert(NoteData { height, note });
+            let key = note.hash().to_bytes();
+
+            if let Some(data_bytes) = self.db.get_cf(cf, &key)? {
+                let mut source = Source::new(&data_bytes);
+                data = KeyData::decode(&mut source)?;
+            }
+
+            data.notes.insert(NoteData { height, note });
+
+            let value = data.encode_to_vec();
+
+            self.write_batch.put_cf(cf, key, value);
         }
 
+        self.write_batch
+            .put_cf(cf, b"last_height", height.to_be_bytes());
+
+        Ok(())
+    }
+
+    pub(crate) fn persist(&self) -> Result<(), Error> {
+        self.db
+            .write(WriteBatch::from_data(self.write_batch.data()))?;
+
```

Comment:
Since you're basically cloning the `write_batch` data, this means that in the next `persist` phase the previous batch is still present and written again (in your case the data is not written multiple times because rocksd overwrite the data based on the key, but this is still a performance issue)

You should ensure that no data is written multiple time, so you have 2 option:
1) Use the trick to encapsulate the `write_batch` into an option so to use the `.take()` method
2) Call `self.write_batch.clear()` after the `.write`